api/src/ipc.c

```diff
@@ -108,6 +108,10 @@ static void ipc_free_allocated_completed_tokens(uint32_t tokens)
 
 static int ipc_waitfor(int (*cond)(uint32_t have, uint32_t expect), uint32_t wait_tokens, uint32_t * done_tokens, uint32_t timeout_ms)
 {
+    /* Check we are not waiting for some unallocated tokens */
+    if ((wait_tokens & *ipc_allocated_tokens()) != wait_tokens) {
```

Comment:
Reading of allocated tokens is not atomic and must be synchronized with the allocation of tokens. Consider using the tokens lock while reading the allocated tokens.